### PR TITLE
[Scheduler] Fair queuing in raylet

### LINKED ISSUE #16973: Improve handling of nested tasks to avoid crashing due to too many workers
When a workload of nested tasks executes, Ray should prioritize launching under-represented task classes. For example, in the following workload:

```
ray.init(num_cpus=2)

@ray.remote
def outer():
   return ray.get(inner.remote())

@ray.remote
def inner():
  return 0

ray.get([f.remote() for _ in range(1000)])
```

Ray should be executing at most 2 concurrent instances of `outer`. Priority should be given to scheduling `inner` tasks. This may be a regression.

cc @wuisawesome @stephanie-wang @anabranch

### PR DESCRIPTION
**Pull request by wuisawesome**
## Why are these changes needed?

This PR introduces fair queuing to the raylet. Fairness is defined in terms of the number of instances of each scheduling class being run at once. 

It also adds FunctionDescriptor to SchedulingClass. 

## Related issue number

Closes #16973

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/core_worker/test/direct_task_transport_test.cc [hunk 1]
```diff
@@ -1007,12 +1007,6 @@ TEST(DirectTaskTransportTest, TestSchedulingKeys) {
                     BuildTaskSpec(resources1, descriptor1),
                     BuildTaskSpec(resources2, descriptor1));
 
-  // Tasks with different function descriptors do not request different worker leases.
```
Comment (wuisawesome) [inline src/ray/core_worker/test/direct_task_transport_test.cc:1010]: Deleted because we no longer actually want this behavior.

File: python/ray/tests/test_scheduling.py [hunk 2]
```diff
@@ -477,6 +477,29 @@ def put():
...
+def test_nested_task(ray_start_regular_shared):
+    @ray.remote
+    def outer():
+        ref = inner.remote()
+        # This sleep is necessary due to a race condition because workers are
+        # blocked/unblocked from the main thread, while tasks are submitted
+        # from a dedicated thread. If the worker is blocked before the inner
+        # task is submitted, the raylet will schedule another instance of
+        # `outer` since there are no `inner`'s queued.
+        time.sleep(0.01)
```
Comment (iycheng) [inline python/ray/tests/test_scheduling.py:489]: If we don't sleep here, will it fail? It introduces the uncertainty for the test.

File: src/ray/raylet/scheduling/fair_scheduling_queue.h [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+using WorkQueueType =
+    std::map<SchedulingClass, std::deque<Work>, internal::SchedulingPriorityComparator>;
+}  // namespace internal
+
+/* using WorkQueueIterator = std::iterator<std::forward_iterator_tag, const
+ * std::pair<const SchedulingClass, std::deque<Work>>>; */
+using WorkQueueIterator = internal::WorkQueueType::iterator;
+using ConstWorkQueueIterator = internal::WorkQueueType::const_iterator;
+
+class FairSchedulingQueue {
```
Comment (iycheng) [inline src/ray/raylet/scheduling/fair_scheduling_queue.h:38]: Please write some comments to explain what's this queue for. Maybe give some examples to explain the behavior under the hood.

File: src/ray/raylet/scheduling/fair_scheduling_queue.h [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+ private:
+  /// A reference to FairSchedulingQueue::active_works_
+  std::unordered_map<SchedulingClass, uint64_t> &priorities_;
+};
+
+using WorkQueueType =
+    std::map<SchedulingClass, std::deque<Work>, internal::SchedulingPriorityComparator>;
+}  // namespace internal
+
+/* using WorkQueueIterator = std::iterator<std::forward_iterator_tag, const
```
Comment (iycheng) [inline src/ray/raylet/scheduling/fair_scheduling_queue.h:33]: remove?

File: src/ray/raylet/scheduling/fair_scheduling_queue.h [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+struct SchedulingPriorityComparator : public std::less<SchedulingClass> {
+ public:
+  explicit SchedulingPriorityComparator(
+      std::unordered_map<SchedulingClass, uint64_t> &priorities);
+
+  bool operator()(const SchedulingClass &lhs, const SchedulingClass &rhs) const;
+
+ private:
+  /// A reference to FairSchedulingQueue::active_works_
+  std::unordered_map<SchedulingClass, uint64_t> &priorities_;
```
Comment (iycheng) [inline src/ray/raylet/scheduling/fair_scheduling_queue.h:26]: should it be a constant? The comparator shouldn't change the member value I think?

File: src/ray/common/task/task_spec.cc [hunk 6]
```diff
@@ -38,7 +40,7 @@ SchedulingClass TaskSpecification::GetSchedulingClass(const ResourceSet &sched_c
                      << " types of tasks seen, this may reduce performance.";
     }
     sched_cls_to_id_[sched_cls] = sched_cls_id;
-    sched_id_to_cls_[sched_cls_id] = sched_cls;
+    sched_id_to_cls_.emplace(sched_cls_id, sched_cls);
```
Comment (iycheng) [inline src/ray/common/task/task_spec.cc:43]: any reason to change this to emplace? If `sched_cls_id` exists, it won't update the value with emplace.

File: src/ray/common/task/task_spec.h [hunk 7]
```diff
@@ -17,7 +17,52 @@ extern "C" {
...
+struct SchedulingClassDescriptor {
+ public:
+  explicit SchedulingClassDescriptor(const ResourceSet resource_set,
+                                     const FunctionDescriptor function_descriptor)
+      : resource_set_(resource_set), function_descriptor_(function_descriptor) {}
+  explicit SchedulingClassDescriptor(const SchedulingClassDescriptor &sched_cls)
+      : resource_set_(sched_cls.resource_set_),
+        function_descriptor_(sched_cls.function_descriptor_) {}
+
+  const std::unordered_map<std::string, double> GetResourceMap() const {
```
Comment (iycheng) [inline src/ray/common/task/task_spec.h:30]: const ref?

File: src/ray/raylet/scheduling/cluster_task_manager.cc [hunk 8]
```diff
@@ -569,21 +566,19 @@ void ClusterTaskManager::FillPendingActorInfo(rpc::GetNodeStatsReply *reply) con
...
-    for (const auto &work_it : work_queue) {
-      Task task = std::get<0>(work_it);
-      if (task.GetTaskSpecification().IsActorCreationTask()) {
-        if (num_reported++ > kMaxPendingActorsToReport) {
-          break;  // Protect the raylet from reporting too much data.
-        }
-        auto ready_task = reply->add_infeasible_tasks();
-        ready_task->CopyFrom(task.GetTaskSpecification().GetMessage());
+
+  std::function<void(const Work &work)> fn = [&num_reported, reply](const Work &work) {
```
Comment (iycheng) [inline src/ray/raylet/scheduling/cluster_task_manager.cc:570]: nit: `auto fn=`

File: src/ray/raylet/scheduling/cluster_task_manager.h [hunk 9]
```diff
@@ -227,7 +223,8 @@ class ClusterTaskManager : public ClusterTaskManagerInterface {
   /// through queues to cancel tasks, etc.
   /// Queue of lease requests that are waiting for resources to become available.
   /// Tasks move from scheduled -> dispatch | waiting.
-  std::unordered_map<SchedulingClass, std::deque<Work>> tasks_to_schedule_;
+  FairSchedulingQueue tasks_to_schedule_;
+  /* std::unordered_map<SchedulingClass, std::deque<Work>> tasks_to_schedule_; */
```
Comment (iycheng) [inline src/ray/raylet/scheduling/cluster_task_manager.h:227]: remove?

File: src/ray/raylet/scheduling/cluster_task_manager.cc [hunk 10]
```diff
@@ -1168,5 +1166,20 @@ ResourceSet ClusterTaskManager::CalcNormalTaskResources() const {
   return total_normal_task_resources;
 }
 
+void ClusterTaskManager::ForAllQueues(std::function<void(const Work &)> &fn) const {
```
Comment (iycheng) [inline src/ray/raylet/scheduling/cluster_task_manager.cc:1169]: This function looks a little bit wired actually. Basically, it wants to iterate all workers in this queue and we have multiple queues there.  Usually boost.range.join will help to chain containers, but here it's nested with chain. I guess it's fine now.

File: src/ray/raylet/scheduling/cluster_task_manager.cc [hunk 11]
```diff
@@ -773,22 +770,22 @@ bool ClusterTaskManager::AnyPendingTasks(Task *exemplar, bool *any_pending,
...
-  for (const auto &shapes_it : boost::join(tasks_to_dispatch_, tasks_to_schedule_)) {
-    auto &work_queue = shapes_it.second;
-    for (const auto &work_it : work_queue) {
-      const auto &task = std::get<0>(work_it);
-      if (task.GetTaskSpecification().IsActorCreationTask()) {
-        *num_pending_actor_creation += 1;
-      } else {
-        *num_pending_tasks += 1;
-      }
+  std::function<void(const Work &)> fn = [exemplar, any_pending,
```
Comment (iycheng) [inline src/ray/raylet/scheduling/cluster_task_manager.cc:773]: Usually how often do we call this function? I feel it might be very expensive in some cases since there is no early return, which means that if there are a lot of tasks, it's going to take a long time to finish.

File: src/ray/raylet/scheduling/fair_scheduling_queue.h [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+  /// Push a new work onto the queue.
+  ///
+  /// \param work Work to be pushed.
+  void Push(const Work &work);
+
+  /// Set the queue for a scheduling class.
+  ///
+  /// \param scheduling_class The scheduling class of the queue.
+  /// \param work All the work for the scheduling class.
+  void PushAll(const SchedulingClass scheduling_class, const std::deque<Work> work_queue);
```
Comment (iycheng) [inline src/ray/raylet/scheduling/fair_scheduling_queue.h:51]: Should std::deque<worker> a const?

File: src/ray/raylet/scheduling/fair_scheduling_queue.cc [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+    : comparator_(active_tasks_), work_queue_(comparator_) {}
+
+void FairSchedulingQueue::Push(const Work &work) {
+  const auto scheduling_class =
+      std::get<0>(work).GetTaskSpecification().GetSchedulingClass();
+  work_queue_[scheduling_class].push_back(work);
+}
+
+void FairSchedulingQueue::PushAll(const SchedulingClass scheduling_class,
+                                  std::deque<Work> works) {
```
Comment (iycheng) [inline src/ray/raylet/scheduling/fair_scheduling_queue.cc:18]: signature not match?

**Review by fishbone — COMMENTED** [review]
Mostly looks fine. But I'm a little bit new to this. Could you share more insights about what's doing here? I can have another look at this one after. Some comments left.

File: python/ray/tests/test_scheduling.py (hunk 2 above)
Comment (wuisawesome) [reply python/ray/tests/test_scheduling.py]: Fixed the race condition.

File: src/ray/raylet/scheduling/fair_scheduling_queue.h (hunk 5 above)
Comment (wuisawesome) [reply src/ray/raylet/scheduling/fair_scheduling_queue.h:26]: i think it shouldn't be constant. i'm adding a comment about how this works.

File: src/ray/common/task/task_spec.cc (hunk 6 above)
Comment (wuisawesome) [reply src/ray/common/task/task_spec.cc:43]: Hmmm I think we need `emplace` because there's no longer a default empty constructor for the SchedulingClassDescriptor. All this logic is inside an if statement that should ensure there isn't an existing key/value.

File: src/ray/common/task/task_spec.h (hunk 7 above)
Comment (wuisawesome) [reply src/ray/common/task/task_spec.h:30]: Hmm can't seem to do that because that would be returning a reference to a stack object. (the actual resource map is actually being constructed in the function, not just returning an instance variable).

File: src/ray/raylet/scheduling/cluster_task_manager.cc (hunk 10 above)
Comment (wuisawesome) [reply src/ray/raylet/scheduling/cluster_task_manager.cc]: Yeah... `boost::join` doesn't seem to work because the types of the iterators are different now. Maybe this is a bit of a noob c++ question, but pls let me know if there's a way to use `boost::join` still.

File: src/ray/raylet/scheduling/cluster_task_manager.cc (hunk 11 above)
Comment (wuisawesome) [reply src/ray/raylet/scheduling/cluster_task_manager.cc]: Changed it to `std::function<bool(const Work &)>` to allow short circuiting.

File: src/ray/raylet/scheduling/fair_scheduling_queue.cc [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+  const auto scheduling_class = task.GetTaskSpecification().GetSchedulingClass();
+
+  auto it = work_queue_.find(scheduling_class);
+  if (it != work_queue_.end()) {
+    auto queue = std::move(it->second);
+    work_queue_.erase(scheduling_class);
+    active_tasks_[scheduling_class]++;
+    work_queue_.emplace(scheduling_class, queue);
+  } else {
+    active_tasks_[scheduling_class]++;
```
Comment (iycheng) [inline src/ray/raylet/scheduling/fair_scheduling_queue.cc:39]: Move this part and the one in the following function into a separate function:

`AdjustPriority(key, int diff)`

File: src/ray/raylet/scheduling/fair_scheduling_queue.h [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+using WorkQueueType =
+    std::map<SchedulingClass, std::deque<Work>, internal::SchedulingPriorityComparator>;
+}  // namespace internal
+
+using WorkQueueIterator = internal::WorkQueueType::iterator;
+using ConstWorkQueueIterator = internal::WorkQueueType::const_iterator;
+
+/// A fair scheduling queue. Fairness is defined in terms of the number of
+/// running tasks of a given scheduling class. For example, assume we push 3
+/// tasks with scheduling classes: {1, 2, 2}. Then we mark a task of class 1 as
```
Comment (iycheng) [inline src/ray/raylet/scheduling/fair_scheduling_queue.h:38]: I feel a little bit hard to follow here, once 1 is marked as running, there are only 2s left, right? so no matter what we'll schedule class 2 I think?

File: src/ray/rpc/node_manager/node_manager_client.h [hunk 16]
```diff
@@ -115,6 +115,9 @@ class NodeManagerWorkerClient
   /// Get gcs server address.
   VOID_RPC_CLIENT_METHOD(NodeManagerService, GetGcsServerAddress, grpc_client_, )
 
+  /// A worker is blocked.
+  VOID_RPC_CLIENT_METHOD(NodeManagerService, DirectCallTaskBlocked, grpc_client_, )
```
Comment (iycheng) [inline src/ray/rpc/node_manager/node_manager_client.h:119]: Could you give me some explanation about why we need this one? I can't link it with the queue here easily.

File: src/ray/raylet/scheduling/fair_scheduling_queue_test.cc [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+  /*
+   * When tasks are queued, make sure we prioritize the shapes that aren't
+   * currently running.
+   */
+  Task task = CreateTask({{ray::kCPU_ResourceLabel, 4}});
+  Task task1 = CreateTask({{ray::kCPU_ResourceLabel, 2}});
+  Task task2 = CreateTask({{ray::kCPU_ResourceLabel, 1}});
+
+  queue_.MarkRunning(task);
+  queue_.MarkRunning(task);
```
Comment (iycheng) [inline src/ray/raylet/scheduling/fair_scheduling_queue_test.cc:35]: typo? marking running twice for the same key?

File: src/ray/raylet/scheduling/fair_scheduling_queue.cc [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+  }
+}
+
+void FairSchedulingQueue::MarkFinished(const Task &task) {
```
Comment (iycheng) [inline src/ray/raylet/scheduling/fair_scheduling_queue.cc:43]: I'm ok with this implementation. But the open question, should accept scheduling_class be good enough?

**Review by fishbone — COMMENTED** [review]
some comments

File: src/ray/raylet/scheduling/fair_scheduling_queue.cc [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+                                  std::deque<Work> works) {
+  // TODO (Alex): We may want to do an O(1) move here if this gets too
+  // inefficient.
+  for (const auto &work : works) {
+    Push(work);
+  }
+}
+
+void FairSchedulingQueue::MarkRunning(const Task &task) {
+  // NOTE: It's unsafe to change `active_works_[scheduling_class]` while
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/fair_scheduling_queue.cc:27]: Can you move this comment to line 35?

File: src/ray/raylet/scheduling/fair_scheduling_queue.cc [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+  } else {
+    active_tasks_[scheduling_class]++;
+  }
+}
+
+void FairSchedulingQueue::MarkFinished(const Task &task) {
+  // NOTE: It's unsafe to change `active_works_[scheduling_class]` while
+  // `scheduling_class` is in `work_queue_` since `works_queue_`'s comparator
+  // relies on `active_works_` and thus may become inconsistent.
+  const auto scheduling_class = task.GetTaskSpecification().GetSchedulingClass();
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/fair_scheduling_queue.cc:47]:
```suggestion
  const auto &scheduling_class = task.GetTaskSpecification().GetSchedulingClass();
```

File: src/ray/raylet/scheduling/fair_scheduling_queue.cc [hunk 21]
```diff
@@ -0,0 +1,113 @@
...
+
+  auto it = work_queue_.find(scheduling_class);
+  if (it != work_queue_.end()) {
+    auto queue = std::move(it->second);
+    work_queue_.erase(scheduling_class);
+    active_tasks_[scheduling_class]--;
+    work_queue_.emplace(scheduling_class, queue);
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/fair_scheduling_queue.cc:54]:
```suggestion
    work_queue_.emplace(scheduling_class, std::move(queue));
```

Comment (rkooo567) [reply src/ray/raylet/scheduling/fair_scheduling_queue.cc:54]: Move again to avoid copy?

File: src/ray/raylet/scheduling/fair_scheduling_queue.h [hunk 22]
```diff
@@ -0,0 +1,118 @@
...
+struct SchedulingPriorityComparator : public std::less<SchedulingClass> {
+ public:
+  explicit SchedulingPriorityComparator(
+      std::unordered_map<SchedulingClass, uint64_t> &priorities);
+
+  bool operator()(const SchedulingClass &lhs, const SchedulingClass &rhs) const;
+
+ private:
+  /// A reference to FairSchedulingQueue::active_works_
+  std::unordered_map<SchedulingClass, uint64_t> &priorities_;
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/fair_scheduling_queue.h:26]: Why don't we just call it `active_works_`?  I think it is okay to expose impl details in this case.

File: src/ray/raylet/scheduling/fair_scheduling_queue_test.cc [hunk 23]
```diff
@@ -0,0 +1,142 @@
...
+ public:
+  FairSchedulingQueue queue_;
+
+  Work TaskToWork(Task &task) { return std::make_tuple(task, nullptr, nullptr); }
+
+  void AssertNoLeaks() {
+    ASSERT_EQ(queue_.active_tasks_.size(), 0);
+    ASSERT_EQ(queue_.work_queue_.size(), 0);
+  }
+};
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/fair_scheduling_queue_test.cc:23]: Can you have a separate test to test `work_queue`?  You can verify

- if exposed methods are working
- if priority orders are calculated correctly.

File: src/ray/raylet/scheduling/fair_scheduling_queue_test.cc (hunk 17 above)
Comment (rkooo567) [reply src/ray/raylet/scheduling/fair_scheduling_queue_test.cc:35]: I think he wanted to simulate two of 4 cpus shapes are running concurrently?

File: src/ray/raylet/scheduling/fair_scheduling_queue_test.cc [hunk 24]
```diff
@@ -0,0 +1,142 @@
...
+  queue_.MarkRunning(task1);
+
+  queue_.Push(TaskToWork(task));
+  queue_.Push(TaskToWork(task1));
+  queue_.Push(TaskToWork(task2));
+
+  std::vector<std::pair<const SchedulingClass, std::deque<Work>>> ordered(queue_.begin(),
+                                                                          queue_.end());
+
+  Task expected[] = {task2, task1, task};
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/fair_scheduling_queue_test.cc:45]: Isn't task1 supposed to run for the first time by FIFO behavior?

File: src/ray/raylet/scheduling/fair_scheduling_queue_test.cc [hunk 25]
```diff
@@ -0,0 +1,142 @@
...
+    const Work &work = deque.front();
+    const Task &t = std::get<0>(work);
+    ASSERT_EQ(t.GetTaskSpecification().TaskId(),
+              expected[i].GetTaskSpecification().TaskId());
+  }
+}
+
+TEST_F(FairSchedulingQueueTest, TestPushAll) {
+  /*
+   * When tasks are queued, make sure we prioritize the shapes that aren't
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/fair_scheduling_queue_test.cc:59]: This explanation doesn't seem right? Also, can you actually test this scenario? `When tasks are queued, make sure we prioritize the shapes that aren't running` with this test ? It looks like we only have a single shape task?

File: src/ray/raylet/scheduling/fair_scheduling_queue_test.cc [hunk 26]
```diff
@@ -0,0 +1,142 @@
...
+  /*
+   * Ensure we dont' leak memory by accidently storing keys after the queue has been
+   * removed.
+   */
+  Task task = CreateTask({{ray::kCPU_ResourceLabel, 4}});
+  queue_.Push(TaskToWork(task));
+  auto it = queue_.begin();
+  queue_.erase(it);
+
+  queue_.MarkRunning(task);
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/fair_scheduling_queue_test.cc:112]: Can you also test without calling `  auto it = queue_.begin();
  queue_.erase(it);`?

**Comment by ericl** [comment]
Seems deprecated by https://github.com/ray-project/ray/pull/17887
